lib/parser/openmp-grammar.h

```diff
@@ -445,73 +416,69 @@ TYPE_PARSER("ATOMIC" >>
-                              Parser<OpenMPDeclareSimdConstruct>{}) /
-                endOmpLine ||
+                              Parser<OpenMPDeclareSimdConstruct>{}) ||
         "DECLARE TARGET" >> construct<OpenMPDeclarativeConstruct>(
                                 construct<OpenMPDeclarativeConstruct>(
-                                    Parser<OpenMPDeclareTargetConstruct>{})) /
-                endOmpLine ||
+                                    Parser<OpenMPDeclareTargetConstruct>{})) ||
         "THREADPRIVATE" >>
             construct<OpenMPDeclarativeConstruct>(
                 construct<OpenMPDeclarativeConstruct::Threadprivate>(
-                    parenthesized(Parser<OmpObjectList>{})) /
-                endOmpLine)))
+                    parenthesized(Parser<OmpObjectList>{}))) /
+                endOmpLine))
 
 // Block Construct
-TYPE_PARSER(construct<OpenMPBlockConstruct>(
-    sourced(Parser<OmpBlockDirective>{}), Parser<OmpClauseList>{} / endOmpLine,
-    block, Parser<OmpEndBlockDirective>{} / endOmpLine))
+TYPE_PARSER(construct<OpenMPBlockConstruct>(Parser<OmpBlockDirective>{},
+    Parser<OmpClauseList>{} / endOmpLine, block,
+    Parser<OmpEndBlockDirective>{}))
+
+// Simple constructs without clauses
+TYPE_PARSER(
+    sourced(construct<OpenMPSimpleConstruct>(first(
+        "BARRIER" >> pure(OpenMPSimpleConstruct::Directive::Barrier),
+        "TASKWAIT" >> pure(OpenMPSimpleConstruct::Directive::Taskwait),
+        "TASKYIELD" >> pure(OpenMPSimpleConstruct::Directive::Taskyield)))) /
+    endOmpLine)
 
 TYPE_PARSER(construct<OpenMPStandaloneConstruct>(
     Parser<OmpStandaloneDirective>{}, Parser<OmpClauseList>{} / endOmpLine))
 
-// OMP BARRIER
-TYPE_PARSER("BARRIER" >> construct<OpenMPBarrierConstruct>() / endOmpLine)
-
-// OMP TASKWAIT
-TYPE_PARSER("TASKWAIT" >> construct<OpenMPTaskwaitConstruct>() / endOmpLine)
-
-// OMP TASKYIELD
-TYPE_PARSER("TASKYIELD" >> construct<OpenMPTaskyieldConstruct>() / endOmpLine)
-
 // OMP SINGLE
-TYPE_PARSER(skipStuffBeforeStatement >> "!$OMP "_sptok >> "END"_tok >>
-    construct<OmpEndSingle>("SINGLE"_tok >> Parser<OmpClauseList>{}))
+TYPE_PARSER(startOmpLine >> "END"_tok >>
+    construct<OmpEndSingle>("SINGLE" >> Parser<OmpClauseList>{}) / endOmpLine)
 
 TYPE_PARSER("SINGLE" >>
-    construct<OpenMPSingleConstruct>(Parser<OmpClauseList>{} / endOmpLine,
-        block, Parser<OmpEndSingle>{} / endOmpLine))
-
-TYPE_PARSER(skipStuffBeforeStatement >> "!$OMP "_sptok >> "END"_tok >>
-    construct<OmpEndWorkshare>("WORKSHARE"_tok))
+    construct<OpenMPSingleConstruct>(
+        OmpConstructDirective("SINGLE"_tok), block, Parser<OmpEndSingle>{}))
 
 // OMP WORKSHARE
-TYPE_PARSER("WORKSHARE" >>
-    construct<OpenMPWorkshareConstruct>(endOmpLine >> block,
-        Parser<OmpEndWorkshare>{} >>
-            maybe(construct<OmpNowait>("NOWAIT"_tok)) / endOmpLine))
+TYPE_PARSER(construct<OpenMPWorkshareConstruct>(
+    verbatim("WORKSHARE"_tok) / endOmpLine, block,
```

Comment:
It fails to get the source location using `verbatim`:
```
!$omp workshare
  a = 1.0
  !$omp end workshare
end
```
OR
```
b = 0.0
  !$omp sections
  !$omp section
  do i=1, 100
     do j=1, 200
        b = b + 1.0
     enddo
  enddo
  !$omp end sections nowait
end
```
Tree looks like:
`ExecutionPartConstruct -> ExecutableConstruct -> OpenMPConstruct -> OmpSection -> Verbatim = ''`